### PR TITLE
Create folders recursively in `rocksdb_open()`

### PR DESCRIPTION
**Pull request by yassernasc**

### CODE REVIEW

**Comment by yassernasc** [comment]
There's some problems with the `libpath` dependency, I had to patch those changes to be able to bypass the compiler errors reproduced on CI:

image

File: src/rocksdb.cc [hunk 1]
```diff
@@ -161,6 +162,87 @@ rocksdb__on_open (uv_work_t *handle) {
...
+  auto req = reinterpret_cast<rocksdb_open_t *>(handle->data);
+
+  rocksdb_fs_mkdir_t *mkdir_req = (rocksdb_fs_mkdir_t *) malloc(sizeof(rocksdb_fs_mkdir_t));
+  mkdir_req->mode = 0777;
+  mkdir_req->path = req->path;
+
+  rocksdb_fs_mkdir_step_t *step = (rocksdb_fs_mkdir_step_t *) malloc(sizeof(rocksdb_fs_mkdir_step_t));
+  step->req = mkdir_req;
+  step->next = NULL;
+  step->handle = req;
```
Comment (kasperisager) [inline src/rocksdb.cc:239]: As per my previous comment, the necessary data should all be defined in the `rocksdb_open_t` struct. A hint is that there's now a constant number (2) of unconditional `malloc()` calls, which is unnecessary.

File: test/open-close.c [hunk 2]
```diff
@@ -43,7 +43,7 @@ main () {
   };
 
   static rocksdb_open_t open;
-  e = rocksdb_open(&db, &open, "test/fixtures/open-close.db", &options, on_open);
+  e = rocksdb_open(&db, &open, "test/fixtures/create/folders/recursively/open-close.db", &options, on_open);
```
Comment (kasperisager) [inline test/open-close.c:46]: Let's add a separate test that specifically tests recursive directory creation.

**Review by kasperisager — CHANGES_REQUESTED** [review]

File: src/rocksdb.cc [hunk 3]
```diff
@@ -161,6 +162,87 @@ rocksdb__on_open (uv_work_t *handle) {
...
+
+  mkdir_req->req.data = (void *) step;
+
+  uv_fs_mkdir(req->db->loop, &mkdir_req->req, mkdir_req->path, mkdir_req->mode, rocksdb__on_recursive_mkdir);
+}
```
[the remark this answers is not in the document]
Comment (yassernasc) [reply src/rocksdb.cc]: That adjustment does the trick? https://github.com/holepunchto/librocksdb/pull/3/commits/7b3e3e7f1c88f7aa87ed6fe28a20da564719ab08

File: src/rocksdb.cc [hunk 4]
```diff
@@ -170,9 +240,11 @@ rocksdb_open (rocksdb_t *db, rocksdb_open_t *req, const char *path, const rocksd
 
   strcpy(req->path, path);
 
+  req->mkdir_next = NULL;
+
   req->worker.data = static_cast<void *>(req);
 
-  return uv_queue_work(db->loop, &req->worker, rocksdb__on_open, rocksdb__on_status<rocksdb_open_t>);
+  return uv_queue_work(db->loop, &req->worker, rocksdb__recursive_mkdir, NULL);
```
Comment (kasperisager) [inline src/rocksdb.cc:247]: This still runs `uv_fs_mkdir()` from the thread pool, which isn't safe. It should run on the same thread that `rocksdb_open()` was called from.

**Review by kasperisager — CHANGES_REQUESTED** [review]

**Review by kasperisager — APPROVED** [review]

**Pull request merged**